srml/evm/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+	/// Otherwise, calculate the amount of gas needed with given `input` and
+	/// `target_gas`. Return `Ok(Some(status, output, gas_used))` if the execution
+	/// is successful. Otherwise return `Ok(Err(_))`.
+	fn execute(
+		address: H160,
+		input: &[u8],
+		target_gas: Option<usize>
+	) -> Option<core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>>;
+}
+
+impl Precompiles for () {
+	fn execute(
+		_address: H160,
+		_input: &[u8],
+		_target_gas: Option<usize>
+	) -> Option<core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>> {
+		None
+	}
+}
+
+/// EVM module trait
+pub trait Trait: system::Trait + timestamp::Trait {
+	/// Calculator for current gas price.
+	type FeeCalculator: FeeCalculator;
+	/// Convert account ID to H160;
+	type ConvertAccountId: ConvertAccountId<Self::AccountId>;
+	/// Currency type for deposit and withdraw.
+	type Currency: Currency<Self::AccountId>;
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+	/// Precompiles associated with this EVM engine.
+	type Precompiles: Precompiles;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		Accounts get(fn accounts) config(): map H160 => Account;
+		AccountStorages: double_map H160, blake2_256(H256) => H256;
+	}
+}
+
+decl_event!(
+	/// EVM events
+	pub enum Event {
+		/// Ethereum events from contracts.
+		Log(Log),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn deposit_balance(origin, value: BalanceOf<T>) -> Result {
+			let sender = ensure_signed(origin)?;
+
+			T::Currency::withdraw(
+				&sender,
+				value,
+				WithdrawReason::Reserve,
+				ExistenceRequirement::KeepAlive,
+			)?;
+
+			let bvalue = U256::from(UniqueSaturatedInto::<u64>::unique_saturated_into(value));
+			let address = T::ConvertAccountId::convert_account_id(sender);
+			Accounts::mutate(&address, |account| {
+				account.balance += bvalue;
+			});
+
+			Ok(())
+		}
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn withdraw_balance(origin, value: BalanceOf<T>) -> Result {
+			let sender = ensure_signed(origin)?;
+			let address = T::ConvertAccountId::convert_account_id(sender.clone());
+			let bvalue = U256::from(UniqueSaturatedInto::<u64>::unique_saturated_into(value));
+
+			if Accounts::get(&address).balance < bvalue {
+				return Err("Not enough balance to withdraw")
+			}
+
+			Accounts::mutate(&address, |account| {
+				account.balance -= bvalue;
+			});
+
+			T::Currency::deposit_creating(&sender, value);
+
+			Ok(())
+		}
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn call(origin, target: H160, input: Vec<u8>, value: U256, gas_limit: u32) -> Result {
+			let sender = ensure_signed(origin)?;
+			let source = T::ConvertAccountId::convert_account_id(sender.clone());
+			let gas_price = T::FeeCalculator::gas_price();
```

Comment:
it will be good if `gas_price` can be calculated from `tip` from balances module somehow

Reply:
Actually, I'd assume that this module should eventually be integrated with the weights system. Otherwise, gas would be just a number that doesn't influence the block weight.

Reply:
It will be the gas limit influence the block weight not gas price?

Reply:
It will be the gas limit influence the block weight not gas price? Gas price should affect the transaction priority.

Reply:
Yes, gas limit should play a similar role to that of the weight. I think gas price should be a derivative from the fee computed for the given gas limit and the price paid for that amount of weight.

Reply:
https://github.com/paritytech/substrate/issues/4007